cabal-install/Distribution/Client/ProjectBuilding.hs

```diff
@@ -0,0 +1,1257 @@
+      return BuildStatusPreExisting
+
+    dryRunPkg (InstallPlan.Configured pkg) depsBuildStatus = do
+      mloc <- checkFetched (pkgSourceLocation pkg)
+      case mloc of
+        Nothing -> return BuildStatusDownload
+
+        Just (LocalUnpackedPackage srcdir) ->
+          -- For the case of a user-managed local dir, irrespective of the
+          -- build style, we build from that directory and put build
+          -- artifacts under the shared dist directory.
+          dryRunLocalPkg pkg depsBuildStatus srcdir
+
+        -- The three tarball cases are handled the same as each other,
+        -- though depending on the build style.
+        Just (LocalTarballPackage    tarball) ->
+          dryRunTarballPkg pkg depsBuildStatus tarball
+
+        Just (RemoteTarballPackage _ tarball) ->
+          dryRunTarballPkg pkg depsBuildStatus tarball
+
+        Just (RepoTarballPackage _ _ tarball) ->
+          dryRunTarballPkg pkg depsBuildStatus tarball
+
+    dryRunPkg (InstallPlan.Processing {}) _ = unexpectedState
+    dryRunPkg (InstallPlan.Installed  {}) _ = unexpectedState
+    dryRunPkg (InstallPlan.Failed     {}) _ = unexpectedState
+
+    unexpectedState = error "rebuildTargetsDryRun: unexpected package state"
+
+    dryRunTarballPkg :: ElaboratedConfiguredPackage
+                     -> ComponentDeps [BuildStatus]
+                     -> FilePath
+                     -> IO BuildStatus
+    dryRunTarballPkg pkg depsBuildStatus tarball =
+      case pkgBuildStyle pkg of
+        BuildAndInstall  -> return (BuildStatusUnpack tarball)
+        BuildInplaceOnly -> do
+          -- TODO: [nice to have] use a proper file monitor rather than this dir exists test
+          exists <- doesDirectoryExist srcdir
+          if exists
+            then dryRunLocalPkg pkg depsBuildStatus srcdir
+            else return (BuildStatusUnpack tarball)
+      where
+        srcdir = distUnpackedSrcDirectory (packageId pkg)
+
+    dryRunLocalPkg :: ElaboratedConfiguredPackage
+                   -> ComponentDeps [BuildStatus]
+                   -> FilePath
+                   -> IO BuildStatus
+    dryRunLocalPkg pkg depsBuildStatus srcdir = do
+        -- Go and do lots of I/O, reading caches and probing files to work out
+        -- if anything has changed
+        change <- checkPackageFileMonitorChanged
+                    packageFileMonitor pkg srcdir depsBuildStatus
+        case change of
+          -- It did change, giving us 'BuildStatusRebuild' info on why
+          Left rebuild ->
+            return (BuildStatusRebuild srcdir rebuild)
+
+          -- No changes, the package is up to date. Use the saved build results.
+          Right (mipkg, buildSuccess) ->
+            return (BuildStatusUpToDate mipkg buildSuccess)
+      where
+        packageFileMonitor =
+          newPackageFileMonitor distDirLayout (packageId pkg)
+
+
+traverseInstallPlan :: forall m ipkg srcpkg iresult ifailure b.
```

Comment:
Needs a comment saying that it feeds in the produced b's of your dependencies. Not really a "traverse" in the traversable sense.